### PR TITLE
Make webdriver client handle headers case insensitively

### PR DESCRIPTION
**Pull request by stephenmcgruer**
Closes https://github.com/web-platform-tests/wpt/pull/25039

### CODE REVIEW

File: tools/webdriver/webdriver/transport.py [hunk 1]
```diff
@@ -10,6 +10,72 @@
 """Implements HTTP transport for the WebDriver wire protocol."""
 
 
+missing = object()
+
+
+class ResponseHeaders(dict):
```
Comment (Hexcles) [inline tools/webdriver/webdriver/transport.py:16]: Shall we inherit from `collections.Mapping` instead? You'd only need to implement `__getitem__`, `__iter__`, `__len__` (and of course `__init__` and `get_list`); the rest will be mixed in for you.

**Review by Hexcles — APPROVED** [review]
LGTM % a nit

Hopefully tests will continue to work...

File: tools/webdriver/webdriver/transport.py (hunk 1 above)
Comment (stephenmcgruer) [reply tools/webdriver/webdriver/transport.py]: Yes, a good idea (done). We can probably drop `get_list` until we have a user, I expect.

**Comment by stephenmcgruer** [comment]
> Hopefully tests will continue to work...

I will be definitely doing trigger runs of everything that uses this client library before merging this.

File: tools/webdriver/webdriver/transport.py [hunk 2]
```diff
@@ -13,67 +14,48 @@
...
-    This class always returns all headers with the same name (separated by
-    commas). It doesn't ensure header types (e.g. binary vs string).
+    This class handles header case-insensitivity, and always returns all
+    headers with the same name (separated by commas). It doesn't ensure header
+    types (e.g. binary vs string).
     """
     def __init__(self, items):
+        self.headers_dict = {}
         for key, value in items:
             key = key.lower()
```
Comment (Hexcles) [inline tools/webdriver/webdriver/transport.py:27]: So we are not preserving the case of keys here, which is probably fine, but we should definitely document it.

File: tools/webdriver/webdriver/transport.py [hunk 3]
```diff
@@ -10,6 +11,53 @@
...
+    """
+    def __init__(self, items):
+        self.headers_dict = {}
+        for key, value in items:
+            key = key.lower()
+            values = []
+            if key in self.headers_dict:
+                values = self.headers_dict[key]
+            values.append(value)
+            self.headers_dict[key] = values
```
Comment (Hexcles) [inline tools/webdriver/webdriver/transport.py:32]: Nit
```suggestion
            if key not in self.headers_dict:
                self.headers_dict[key] = []
            self.headers_dict[key].append(value)
```

File: tools/webdriver/webdriver/transport.py [hunk 4]
```diff
@@ -10,6 +11,53 @@
...
+
+    def __getitem__(self, key):
+        """Get all headers of a certain (case-insensitive) name. If there is
+        more than one, the values are returned comma separated"""
+        values = self.headers_dict[key.lower()]
+        if len(values) == 1:
+            return values[0]
+        else:
+            return b", ".join(values)
```
Comment (Hexcles) [inline tools/webdriver/webdriver/transport.py:41]: The docstring says the class doesn't ensure types, but we are assuming bytes here?

File: tools/webdriver/webdriver/transport.py (hunk 2 above)
Comment (stephenmcgruer) [reply tools/webdriver/webdriver/transport.py:33]: Done.

File: tools/webdriver/webdriver/transport.py (hunk 4 above)
Comment (stephenmcgruer) [reply tools/webdriver/webdriver/transport.py:45]: Eugh, good catch. I was hoping to get away with not caring about this for now, but having this here will mean different Py2/Py3 behavior.

We could either drop the support for multiple header values (and return to last-one-wins), or we could return an array here (I think that would give us pretty surprising behavior however), or possibly some other idea I haven't thought of yet. Any preference?

Comment (Hexcles) [reply tools/webdriver/webdriver/transport.py:45]: We could use a native string here, too. I believe the values are always native strings coming from the http library.

Comment (stephenmcgruer) [reply tools/webdriver/webdriver/transport.py]: SGTM, done.

**Comment by stephenmcgruer** [comment]
Ok, I've rebased this and triggered a Chrome Dev run to start with.

**Comment by stephenmcgruer** [comment]
So I forgot I hadn't landed this 😆 .

The stability checks timed out (unsurprising), the Firefox and Safari wpt.fyi results are the known flaky webdriver tests, Chrome wpt.fyi results look good.

The Chrome Dev diff is https://wpt.fyi/results/?diff&filter=ADC&run_id=700170006&run_id=681230006, I'm digging into it now.

cc @ziransun as an FYI.

**Comment by stephenmcgruer** [comment]
> The Chrome Dev diff is https://wpt.fyi/results/?diff&filter=ADC&run_id=700170006&run_id=681230006, I'm digging into it now.

This looks good; all the diffs appear to be tests that are already flaky.

I'll give this change 24 hours to give a chance for folks to raise concerns, then will merge based on LGTM from @Hexcles

**Pull request merged**